File: plugin/src/main/java/broccolai/tickets/commands/TicketsCommand.java
package broccolai.tickets.commands;

import broccolai.corn.core.Lists;
import broccolai.tickets.commands.arguments.TicketArgument;
import broccolai.tickets.configuration.Config;
import broccolai.tickets.exceptions.PureException;
import broccolai.tickets.interactions.NotificationManager;
import broccolai.tickets.locale.MessageNames;
import broccolai.tickets.locale.Messages;
import broccolai.tickets.storage.TimeAmount;
import broccolai.tickets.ticket.Ticket;
import broccolai.tickets.ticket.TicketManager;
import broccolai.tickets.ticket.TicketStats;
import broccolai.tickets.ticket.TicketStatus;
import broccolai.tickets.user.PlayerSoul;
import broccolai.tickets.user.Soul;
import broccolai.tickets.user.UserManager;
import broccolai.tickets.utilities.Constants;
import broccolai.tickets.utilities.ReplacementUtilities;
import broccolai.tickets.utilities.UserUtilities;
import cloud.commandframework.Command;
import cloud.commandframework.Description;
import cloud.commandframework.arguments.CommandArgument;
import cloud.commandframework.arguments.standard.BooleanArgument;
import cloud.commandframework.arguments.standard.EnumArgument;
import cloud.commandframework.arguments.standard.StringArgument;
import cloud.commandframework.bukkit.parsers.OfflinePlayerArgument;
import cloud.commandframework.context.CommandContext;
import com.google.common.collect.ImmutableMap;
import io.papermc.lib.PaperLib;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.checkerframework.checker.nullness.qual.NonNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

public final class TicketsCommand extends BaseCommand {

    private final NotificationManager notificationManager;
    private final TicketManager ticketManager;

    /**
     * Create a new Tickets Command.
     *
     * @param manager             Command Manager
     * @param config              Config Instance
     * @param userManager         User Manager
     * @param notificationManager Notification Manager
     * @param ticketManager       Ticket Manager
     */
    public TicketsCommand(
            final @NonNull CommandManager manager, final @NonNull Config config, final @NonNull UserManager userManager,
            final @NonNull NotificationManager notificationManager, final @NonNull TicketManager ticketManager
    ) {
        this.notificationManager = notificationManager;
        this.ticketManager = ticketManager;

        final Command.Builder<Soul> builder = manager.commandBuilder("tickets", "tis");
        final CommandArgument<Soul, OfflinePlayer> targetArgument = manager.argumentBuilder(OfflinePlayer.class, "target")
//                .withSuggestionsProvider((context, input) -> userManager.getNames())
                .build();

        manager.command(builder.literal(
                config.getAliasShow().getFirst(),
                Description.of("Show a ticket"),
                config.getAliasShow().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".show")
                .argument(targetArgument.copy())
                .argument(TicketArgument.of(false, false))
                .handler(c -> processShow(c.getSender(), c.get("ticket"))));

        manager.command(builder.literal(
                config.getAliasPick().getFirst(),
                Description.of("Pick a ticket"),
                config.getAliasPick().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".pick")
                .argument(targetArgument.copy())
                .argument(TicketArgument.of(false, false, TicketStatus.OPEN))
                .handler(this::processPick)
                .build());

        manager.command(builder.literal(
                config.getAliasAssign().getFirst(),
                Description.of("Assign a ticket"),
                config.getAliasAssign().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".assign")
                .argument(OfflinePlayerArgument.of("staff"))
                .argument(targetArgument.copy())
                .argument(TicketArgument.of(false, false, TicketStatus.OPEN))
                .handler(this::processAssign)
                .build());

        manager.command(builder.literal(
                config.getAliasDone().getFirst(),
                Description.of("Done-mark a ticket"),
                config.getAliasDone().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".done")
                .argument(targetArgument.copy())
                .argument(TicketArgument.of(false, false, TicketStatus.OPEN, TicketStatus.PICKED))
                .handler(this::processDone)
                .build());

        manager.command(builder.literal(
                config.getAliasYield().getFirst(),
                Description.of("Yield a ticket"),
                config.getAliasYield().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".yield")
                .argument(targetArgument.copy())
                .argument(TicketArgument.of(false, false, TicketStatus.PICKED))
                .handler(this::processYield)
                .build());

        manager.command(builder.literal(
                config.getAliasNote().getFirst(),
                Description.of("Add a note to a ticket"),
                config.getAliasNote().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".note")
                .argument(targetArgument.copy())
                .argument(TicketArgument.of(true, false))
                .argument(StringArgument.of("message"))
                .handler(this::processNote)
                .build());

        manager.command(builder.literal(
                config.getAliasReopen().getFirst(),
                Description.of("Reopen a ticket"),
                config.getAliasReopen().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".reopen")
                .argument(targetArgument.copy())
                .argument(TicketArgument.of(false, false, TicketStatus.CLOSED))
                .handler(this::processReopen)
                .build());

        manager.command(builder.literal(
                config.getAliasTeleport().getFirst(),
                Description.of("Teleport to a tickets creation location"),
                config.getAliasTeleport().getSecond()
        )
                .senderType(PlayerSoul.class)
                .permission(Constants.STAFF_PERMISSION + ".teleport")
                .argument(targetArgument.copy())
                .argument(TicketArgument.of(false, false, TicketStatus.OPEN, TicketStatus.PICKED, TicketStatus.CLOSED))
                .handler(this::processTeleport)
                .build());

        manager.command(builder.literal(
                config.getAliasLog().getFirst(),
                Description.of("View a tickets log"),
                config.getAliasLog().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".log")
                .argument(targetArgument.copy())
                .argument(TicketArgument.of(false, false, TicketStatus.OPEN, TicketStatus.PICKED, TicketStatus.CLOSED))
                .handler(c -> processLog(c.getSender(), c.get("ticket")))
                .build());

        manager.command(builder.literal(
                config.getAliasList().getFirst(),
                Description.of("List tickets"),
                config.getAliasList().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".list")
                .flag(manager.flagBuilder("status")
                        .withArgument(EnumArgument.of(TicketStatus.class, "status")))
                .flag(manager.flagBuilder("player")
                        .withArgument(OfflinePlayerArgument.of("player")))
                .flag(manager.flagBuilder("onlineOnly")
                        .withArgument(BooleanArgument.of("onlineOnly")))
                .handler(this::processList));

        manager.command(builder.literal(
                config.getAliasStatus().getFirst(),
                Description.of("View amount of tickets in"),
                config.getAliasStatus().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".status")
                .argument(targetArgument.copy())
                .handler(this::processStatus)
                .build());

        manager.command(builder.literal(
                config.getAliasHighscore().getFirst(),
                Description.of("View highscores of ticket completions"),
                config.getAliasHighscore().getSecond()
        )
                .permission(Constants.STAFF_PERMISSION + ".highscore")
                .argument(EnumArgument.optional(TimeAmount.class, "amount"))
                .handler(this::processHighscore)
                .build());
    }

    private void processPick(final @NonNull CommandContext<Soul> c) {
        Soul soul = c.getSender();
        Ticket ticket = c.get("ticket");

        try {
            ticket.pick(soul.getUniqueId());
            notificationManager.send(soul, ticket.getPlayerUUID(), MessageNames.PICK_TICKET, ticket);
        } catch (PureException e) {
            notificationManager.handleException(soul, e);
        }
    }

    private void processAssign(final @NonNull CommandContext<Soul> c) {
        Soul soul = c.getSender();
        Ticket ticket = c.get("ticket");
        OfflinePlayer staff = c.get("staff");

        try {
            ticket.pick(staff.getUniqueId());
            notificationManager.send(soul, staff.getUniqueId(), MessageNames.ASSIGN_TICKET, ticket);
        } catch (PureException e) {
            notificationManager.handleException(soul, e);
        }
    }

    private void processDone(final @NonNull CommandContext<Soul> c) {
        Soul soul = c.getSender();
        Ticket ticket = c.get("ticket");

        try {
            ticket.done(soul.getUniqueId());
            notificationManager.send(soul, ticket.getPlayerUUID(), MessageNames.DONE_TICKET, ticket);
        } catch (PureException e) {
            notificationManager.handleException(soul, e);
        }
    }

    private void processYield(final @NonNull CommandContext<Soul> c) {
        Soul soul = c.getSender();
        Ticket ticket = c.get("ticket");

        try {
            ticket.yield(soul.getUniqueId());
            notificationManager.send(soul, ticket.getPlayerUUID(), MessageNames.YIELD_TICKET, ticket);
        } catch (PureException e) {
            notificationManager.handleException(soul, e);
        }
    }

    private void processNote(final @NonNull CommandContext<Soul> c) {
        Soul soul = c.getSender();
        Ticket ticket = c.get("ticket");

        ticket.note(soul.getUniqueId(), c.get("message"));
        notificationManager.send(soul, ticket.getPlayerUUID(), MessageNames.NOTE_TICKET, ticket);
    }

    private void processReopen(final @NonNull CommandContext<Soul> c) {
        Soul soul = c.getSender();
        Ticket ticket = c.get("ticket");

        try {
            ticket.reopen(soul.getUniqueId());
            notificationManager.send(soul, ticket.getPlayerUUID(), MessageNames.REOPEN_TICKET, ticket);
        } catch (PureException e) {
            notificationManager.handleException(soul, e);
        }
    }

    private void processTeleport(final @NonNull CommandContext<Soul> c) {
        PlayerSoul soul = (PlayerSoul) c.getSender();
        Ticket ticket = c.get("ticket");

        notificationManager.send(soul, ticket.getPlayerUUID(), MessageNames.TELEPORT_TICKET, ticket);
        PaperLib.teleportAsync(soul.asPlayer(), ticket.getLocation());
    }

    private void processList(final @NonNull CommandContext<Soul> c) {
        final Soul soul = c.getSender();
        final OfflinePlayer player = c.flags().getValue("player", null);
        final Boolean onlineOnly = c.flags().getValue("onlineOnly", false);
        final TicketStatus[] statuses = statusesFromFlags(c.flags());

        if (player != null) {
            soul.message(Messages.TITLES__SPECIFIC_TICKETS, "player", player.getName());

            ticketManager.getTickets(player.getUniqueId(), statuses).forEach(ticket -> {
                String[] replacements = ReplacementUtilities.ticketReplacements(ticket);
                soul.message(Messages.GENERAL__LIST_FORMAT, replacements);
            });
            return;
        }

        soul.message(Messages.TITLES__ALL_TICKETS);

        // todo: ugly
        Set<Map.Entry<UUID, List<Ticket>>> unsortedTickets = Lists
                .group(ticketManager.getTickets(statuses), Ticket::getPlayerUUID)
                .entrySet();
        List<Map.Entry<UUID, List<Ticket>>> sortedTickets = new ArrayList<>(unsortedTickets);
        sortedTickets.sort((t1, t2) -> {
            Integer boxed = t1.getValue().get(0).getId();
            return boxed.compareTo(t2.getValue().get(0).getId());
        });

        ImmutableMap.copyOf(sortedTickets).forEach((uuid, tickets) -> {
            OfflinePlayer offlinePlayer = Bukkit.getOfflinePlayer(uuid);

            //noinspection ConstantConditions
            if (onlineOnly && !offlinePlayer.isOnline()) {
                return;
            }

            soul.message(Messages.GENERAL__LIST_HEADER_FORMAT, "name", UserUtilities.nameFromUUID(uuid));

            tickets.forEach(ticket -> {
                String[] replacements = ReplacementUtilities.ticketReplacements(ticket);
                soul.message(Messages.GENERAL__LIST_FORMAT, replacements);
            });
        });
    }

    private void processStatus(final @NonNull CommandContext<Soul> c) {
        Soul soul = c.getSender();
        OfflinePlayer target = c.getOrDefault("target", null);

        if (target != null) {
            soul.message(Messages.TITLES__SPECIFIC_TICKETS, "player", target.getName());
        } else {
            soul.message(Messages.TITLES__TICKET_STATUS);
        }

        TicketStats data;

        if (target != null) {
            data = ticketManager.getStats(target.getUniqueId());
        } else {
            data = ticketManager.getStats();
        }

        data.forEach((status, amount) -> {
            if (amount != 0) {
                soul.message(amount.toString() + " " + status.name().toLowerCase());
            }
        });
    }

    private void processHighscore(final @NonNull CommandContext<Soul> c) {
        Soul soul = c.getSender();
        TimeAmount amount = c.get("amount");

        Map<UUID, Integer> highscores = ticketManager.getHighscores(amount);
        soul.message(Messages.TITLES__HIGHSCORES);

        highscores.forEach((uuid, number) ->
                soul.message(Messages.GENERAL__HS_FORMAT, "target", UserUtilities.nameFromUUID(uuid), "amount", number.toString())
        );
    }

}

File: src/main/kotlin/co/uk/magmo/puretickets/commands/TicketsCommand.kt
package co.uk.magmo.puretickets.commands

import co.aikar.commands.annotation.*
import co.uk.magmo.puretickets.locale.MessageNames
import co.uk.magmo.puretickets.locale.Messages
import co.uk.magmo.puretickets.ticket.TicketStatus
import co.uk.magmo.puretickets.utils.*
import com.okkero.skedule.SynchronizationContext
import org.bukkit.ChatColor
import org.bukkit.OfflinePlayer
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

@CommandAlias("tickets|tis")
@CommandPermission(Constants.STAFF_PERMISSION)
class TicketsCommand : PureBaseCommand() {
    @Subcommand("%show")
    @CommandCompletion("@TicketHolders @TargetIds")
    @CommandPermission(Constants.STAFF_PERMISSION + ".show")
    @Description("Show a ticket")
    @Syntax("<Player> [Index]")
    fun onShow(sender: CommandSender, offlinePlayer: OfflinePlayer, @Optional index: Int?) {
        val information = generateId(offlinePlayer, index)

        processShowCommand(currentCommandIssuer, information)
    }

    @Subcommand("%pick")
    @CommandCompletion("@TicketHolders:status=OPEN @TargetIds")
    @CommandPermission(Constants.STAFF_PERMISSION + ".pick")
    @Description("Pick a ticket")
    @Syntax("<Player> [Index]")
    fun onPick(sender: CommandSender, offlinePlayer: OfflinePlayer, @Optional index: Int?) {
        val id = generateId(offlinePlayer, index, TicketStatus.OPEN, TicketStatus.PICKED)

        taskManager {
            val ticket = ticketManager.pick(sender.asUUID(), id) ?: return@taskManager

            notificationManager.send(sender, offlinePlayer.uniqueId, MessageNames.PICK_TICKET, ticket) { fields ->
                fields["PICKER"] = sender.name
            }
        }
    }

    @Subcommand("%assign")
    @CommandCompletion("@Players @TicketHolders:status=OPEN @TargetIds:parameter=2")
    @CommandPermission(Constants.STAFF_PERMISSION + ".assign")
    @Description("Assign a ticket to a staff member")
    @Syntax("<TargetPlayer> <Player> [Index]")
    fun onAssign(sender: CommandSender, target: OfflinePlayer, offlinePlayer: OfflinePlayer, @Optional index: Int?) {
        val id = generateId(offlinePlayer, index, TicketStatus.OPEN, TicketStatus.PICKED)

        taskManager {
            val ticket = ticketManager.pick(target.uniqueId, id) ?: return@taskManager

            notificationManager.send(sender, target.uniqueId, MessageNames.ASSIGN_TICKET, ticket) { fields ->
                fields["ASSIGNER"] = sender.name
                fields["ASSIGNEE"] = target.name ?: ""
            }
        }
    }

    @Subcommand("%done")
    @CommandCompletion("@TicketHolders:status=PICK @TargetIds")
    @CommandPermission(Constants.STAFF_PERMISSION + ".done")
    @Description("Done-mark a ticket")
    @Syntax("<Player> [Index]")
    fun onDone(sender: CommandSender, offlinePlayer: OfflinePlayer, @Optional index: Int?) {
        val id = generateId(offlinePlayer, index, TicketStatus.OPEN, TicketStatus.PICKED)

        taskManager {
            val ticket = ticketManager.done(sender.asUUID(), id) ?: return@taskManager

            notificationManager.send(sender, offlinePlayer.uniqueId, MessageNames.DONE_TICKET, ticket)
        }
    }

    @Subcommand("%yield")
    @CommandCompletion("@TicketHolders:status=PICK @TargetIds")
    @CommandPermission(Constants.STAFF_PERMISSION + ".yield")
    @Description("Yield a ticket")
    @Syntax("<Player> [Index]")
    fun onYield(sender: CommandSender, offlinePlayer: OfflinePlayer, @Optional index: Int?) {
        val id = generateId(offlinePlayer, index, TicketStatus.PICKED)

        taskManager {
            val ticket = ticketManager.yield(sender.asUUID(), id) ?: return@taskManager

            notificationManager.send(sender, offlinePlayer.uniqueId, MessageNames.YIELD_TICKET, ticket)
        }
    }

    @Subcommand("%note")
    @CommandCompletion("@TicketHolders @TargetIds")
    @CommandPermission(Constants.STAFF_PERMISSION + ".note")
    @Description("Make a note on a ticket")
    @Syntax("<Player> <Index> <Message>")
    fun onNote(sender: CommandSender, offlinePlayer: OfflinePlayer, index: Int, message: String) {
        val id = generateId(offlinePlayer, index)

        taskManager {
            val ticket = ticketManager.note(sender.asUUID(), id, message) ?: return@taskManager

            notificationManager.send(sender, offlinePlayer.uniqueId, MessageNames.NOTE_TICKET, ticket) { fields ->
                fields["NOTE"] = message
            }
        }
    }

    @Subcommand("%reopen")
    @CommandCompletion("@TicketHolders:status=CLOSED @TargetIds")
    @CommandPermission(Constants.STAFF_PERMISSION + ".reopen")
    @Description("Reopen a ticket")
    @Syntax("<Player> [Index]")
    fun onReopen(sender: CommandSender, offlinePlayer: OfflinePlayer, @Optional index: Int?) {
        val id = generateId(offlinePlayer, index, TicketStatus.CLOSED)

        taskManager {
            val ticket = ticketManager.reopen(sender.asUUID(), id) ?: return@taskManager

            notificationManager.send(sender, offlinePlayer.uniqueId, MessageNames.REOPEN_TICKET, ticket)
        }
    }

    @Subcommand("%teleport")
    @CommandCompletion("@TicketHolders @TargetIds:parameter=1")
    @CommandPermission(Constants.STAFF_PERMISSION + ".teleport")
    @Description("Teleport to a ticket creation location")
    @Syntax("<Player> [Index]")
    fun onTeleport(player: Player, offlinePlayer: OfflinePlayer, @Optional index: Int?) {
        val id = generateId(offlinePlayer, index)

        taskManager {
            val ticket = ticketManager[id] ?: return@taskManager
            notificationManager.send(player, offlinePlayer.uniqueId, MessageNames.TELEPORT_TICKET, ticket)

            switchContext(SynchronizationContext.SYNC)
            player.teleport(ticket.location)
        }
    }

    @Subcommand("%log")
    @CommandCompletion("@TicketHolders @TargetIds")
    @CommandPermission(Constants.STAFF_PERMISSION + ".log")
    @Description("Log tickets messages")
    @Syntax("<Player> [Index]")
    fun onLog(sender: CommandSender, offlinePlayer: OfflinePlayer, @Optional index: Int?) {
        val id = generateId(offlinePlayer, index)

        processLogCommand(currentCommandIssuer, id)
    }

    @Subcommand("%list")
    @CommandCompletion("@TicketHolders @TicketStatus")
    @CommandPermission(Constants.STAFF_PERMISSION + ".list")
    @Description("List all tickets")
    @Syntax("[Player]")
    fun onList(sender: CommandSender, @Optional offlinePlayer: OfflinePlayer?, @Optional status: TicketStatus?) {
        val issuer = currentCommandIssuer

        taskManager {
            if (offlinePlayer != null) {
                var tickets = ticketManager[offlinePlayer.uniqueId]
                if (status != null) tickets = tickets.filter { ticket -> ticket.status == status }

                issuer.sendInfo(Messages.TITLES__SPECIFIC_TICKETS, "%player%", offlinePlayer.name!!)

                tickets.forEach { ticket ->
                    val replacements = Utils.ticketReplacements(ticket)

                    issuer.sendInfo(Messages.GENERAL__LIST_FORMAT, *replacements)
                }
            } else {
                issuer.sendInfo(Messages.TITLES__ALL_TICKETS)

                ticketManager.all().groupBy { it.playerUUID }.forEach { (uuid, tickets) ->
                    sender.sendMessage(ChatColor.GREEN.toString() + uuid.asName())

                    tickets.forEach { ticket ->
                        val replacements = Utils.ticketReplacements(ticket)

                        issuer.sendInfo(Messages.GENERAL__LIST_FORMAT, *replacements)
                    }
                }
            }
        }
    }

    @Subcommand("%status")
    @CommandCompletion("@TicketHolders")
    @CommandPermission(Constants.STAFF_PERMISSION + ".status")
    @Description("View amount of tickets in")
    @Syntax("[Player]")
    fun onStatus(sender: CommandSender, @Optional offlinePlayer: OfflinePlayer?) {
        val issuer = currentCommandIssuer

        taskManager {
            if (offlinePlayer != null) {
                issuer.sendInfo(Messages.TITLES__SPECIFIC_STATUS, "%player%", offlinePlayer.name!!)
            } else {
                issuer.sendInfo(Messages.TITLES__TICKET_STATUS)
            }

            val data = ticketManager.stats(offlinePlayer?.uniqueId)

            data.forEach { (status, amount) ->
                if (amount != 0) sender.sendMessage(amount.toString() + " " + status.name.toLowerCase())
            }
        }
    }
}